### PR TITLE
loader: simplify template cache invalidation

### PR DESCRIPTION
**Pull request by lmb**
elf: fix ELF.Write for files obtained from NewELF

    Trying to call ELF.Write on a file obtained from NewELF currently fails with
    EINVAL, since ELF.Write expects ELF.file to be non-nil.

    Fix the code so that it works with NewELF as well.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: remove NewObjectCache

    There are no external users of the object cache, remove the public 
    constructor.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: simplify template cache invalidation

    The loader tries to avoid recompiling endpoint programs if possible by 
    caching them on disk, based on their configuration. To allow reuse it 
    formats the configuration as a header file and hashes that. The hash is then
    used as a key for the cache.

    It also tries to be resilient to something or someone deleting a cached 
    program by creating an inotify watcher on the state directory. This has 
    several drawbacks:

    * inotify watches are a global resource and can therefore run out
    * Eviction of deleted cache objects is asynchronous. The loader may
     therefore intermittently try to use a non-existing cached file.

    Rewrite the cache to just fall back to recompilation if opening the cached
    object from disk fails. This allows getting rid of all the inotify code and
    is functionally equivalent.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

serializer: remove package

    There are no users left in the code base, remove the package.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

### CODE REVIEW

**Review by bimmlerd — APPROVED** [review]
CODEOWNERS lgtm (did not look at the rest)

**Comment by danehans** [comment]
@lmb the fix references a downstream issue (https://github.com/isovalent/cilium/issues/1474). Do you mind creating and linking an upstream issue so I can better understand the reason for the PR?

**Comment by ti-mo** [comment]
@danehans We've been wanting to shorten the compiler + loader pipeline for a while now (see https://github.com/cilium/cilium/issues/25004, for example), but this particular defect came up during internal chaos testing:

The [fsnotify documentation](https://github.com/fsnotify/fsnotify/blob/main/README.md#linux) states:

> When a file is removed a REMOVE event won't be emitted until all file descriptors are closed; it will emit a CHMOD instead […]

The agent doesn't hold open file descriptors to the files in the cache. It also ignores all fsnotify events other than REMOVE. Therefore, if we hold a file descriptor open from another process and then delete the file, the REMOVE event will never be triggered, and the cache won't see the file as deleted. However it will still attempt to read from the file.

Also, as mentioned in the commit messages in this PR, inotify handles are a global resource, and global node configuration can interfere with this. As such, to reduce the amount of moving parts and reduce the potential for failures to occur, we're no longer going to rely on inotify.

File: pkg/datapath/loader/cache.go [hunk 1]
```diff
@@ -166,98 +144,72 @@ func (o *objectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
...
+		prog = hostEpProg
 	}
-	objectPath := filepath.Join(templatePath, epObj)
 
-	if err := os.MkdirAll(templatePath, defaults.StateDirRights); err != nil {
-		return &os.PathError{
+	objectPath := prog.AbsoluteOutput(dir)
+
+	if err := os.MkdirAll(dir.Output, defaults.StateDirRights); err != nil {
+		return "", &os.PathError{
```
Comment (ti-mo) [inline pkg/datapath/loader/cache.go]: Note: this code far predates `%w` and I don't see any reason to be strict about returning `os.PathError`s (the caller just calls `log.WithError()` on this.. Could we cut this down further by simply wrapping?

Same for all following error returns in this function.

File: pkg/datapath/loader/cache.go [hunk 2]
```diff
@@ -159,98 +144,72 @@ func (o *objectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
...
-		fq = serializer.NewFunctionQueue()
-		o.compileQueue[hash] = fq
-	}
-	return fq, compiled
-}
-
-func (o *objectCache) lookup(hash string) (string, bool) {
+// serialize access to an abitrary key.
+//
+// Returns true if this was the first access to the
```
Comment (ti-mo) [inline pkg/datapath/loader/cache.go]: Incomplete comment.

**Review by ti-mo — APPROVED** [review]
Awesome, thanks! I'd like a code walk of this once you're back, I don't have enough understanding of the surrounding system, so I can't give an in-depth review off hand aside from the odd nit.

**Comment by lmb** [comment]
> I don't have enough understanding of the surrounding system, so I can't give an in-depth review off hand aside from the odd nit

Good thing we are early in the cycle :sweat_smile:

**Review by nathanjsweet — APPROVED** [review]
LGTM

**Pull request merged**